src/libmongoc/tests/unified/operation.c

```diff
@@ -0,0 +1,2261 @@
+   _mongoc_array_destroy (&all_bytes);
+   bson_val_destroy (val);
+   mongoc_stream_destroy (stream);
+   return ret;
+}
+
+static bool
+operation_upload (test_t *test,
+                  operation_t *op,
+                  result_t *result,
+                  bson_error_t *error)
+{
+   bool ret = false;
+   mongoc_gridfs_bucket_t *bucket = NULL;
+   bson_parser_t *bp = NULL;
+   bson_t *source = NULL;
+   char *filename = NULL;
+   mongoc_stream_t *stream = NULL;
+   bson_value_t file_id;
+   bson_error_t op_error = {0};
+   bson_val_t *val = NULL;
+
+   bp = bson_parser_new ();
+   bson_parser_allow_extra (bp, true);
+   bson_parser_doc (bp, "source", &source);
+   bson_parser_utf8 (bp, "filename", &filename);
+   if (!bson_parser_parse (bp, op->arguments, error)) {
+      goto done;
+   }
+
+   bucket = entity_map_get_bucket (test->entity_map, op->object, error);
+   if (!bucket) {
+      goto done;
+   }
+
+   stream = mongoc_gridfs_bucket_open_upload_stream (
+      bucket, filename, bson_parser_get_extra (bp), &file_id, &op_error);
+
+   if (stream) {
+      ssize_t total_written = 0;
+      uint8_t *source_bytes;
+      uint32_t source_bytes_len;
+      bson_iter_t iter;
+
+      if (!bson_iter_init_find (&iter, source, "$$hexBytes") ||
+          !BSON_ITER_HOLDS_UTF8 (&iter)) {
+         test_set_error (error, "$$hexBytes not found in source data");
+         goto done;
+      }
+
+      source_bytes =
+         hex_to_bin (bson_iter_utf8 (&iter, NULL), &source_bytes_len);
+      while (total_written < source_bytes_len) {
+         ssize_t bytes_written = 0;
+
+         bytes_written =
+            mongoc_stream_write (stream,
+                                 source_bytes,
+                                 (size_t) (source_bytes_len - total_written),
+                                 0);
+         if (bytes_written < 0) {
+            break;
+         }
+         total_written += bytes_written;
+      }
+      mongoc_gridfs_bucket_stream_error (stream, &op_error);
+      bson_free (source_bytes);
+   }
+
+   val = bson_val_from_value (&file_id);
+   result_from_val_and_reply (result, val, NULL, &op_error);
+
+   ret = true;
+done:
+   bson_parser_destroy_with_parsed_fields (bp);
+   mongoc_stream_destroy (stream);
+   bson_val_destroy (val);
+   return ret;
+}
+
+static bool
+assert_session_dirty_helper (test_t *test,
+                             operation_t *op,
+                             result_t *result,
+                             bool check_dirty,
+                             bson_error_t *error)
+{
+   bool ret = false;
+
+   if (!op->session) {
+      test_set_error (error, "%s", "session unset");
+      goto done;
+   }
+
+   if (check_dirty) {
+      if (check_dirty != mongoc_client_session_get_dirty (op->session)) {
+         test_set_error (error,
+                         "expected session to%s be dirty but was not",
+                         check_dirty ? "" : "not");
```

Comment:
If `check_dirty == false` then this line would not be reached because of the check on line 1489. Maybe delete the check on 1489?